aggregator/entry.go

```diff
@@ -315,170 +339,179 @@ func (e *Entry) recordLastAccessed(currTime time.Time) {
 	atomic.StoreInt64(&e.lastAccessNanos, currTime.UnixNano())
 }
 
-// NB(xichen): we assume the policies are sorted by their cutover times
+// NB(xichen): we assume the metadatas are sorted by their cutover times
 // in ascending order.
-func (e *Entry) activeStagedPoliciesWithLock(
-	pl policy.PoliciesList,
+func (e *Entry) activeStagedMetadataWithLock(
 	t time.Time,
-) (policy.StagedPolicies, error) {
-	// If we have no policy to apply, simply bail.
-	if len(pl) == 0 {
-		e.metrics.emptyPoliciesList.Inc(1)
-		return policy.DefaultStagedPolicies, errEmptyPoliciesList
+	metadatas metadata.StagedMetadatas,
+) (metadata.StagedMetadata, error) {
+	// If we have no metadata to apply, simply bail.
+	if len(metadatas) == 0 {
+		e.metrics.emptyMetadatas.Inc(1)
+		return metadata.DefaultStagedMetadata, errEmptyMetadatas
 	}
 	timeNanos := t.UnixNano()
-	for idx := len(pl) - 1; idx >= 0; idx-- {
-		if pl[idx].CutoverNanos <= timeNanos {
-			return pl[idx], nil
+	for idx := len(metadatas) - 1; idx >= 0; idx-- {
+		if metadatas[idx].CutoverNanos <= timeNanos {
+			return metadatas[idx], nil
 		}
 	}
-	return pl[0], nil
+	e.metrics.noApplicableMetadata.Inc(1)
+	return metadata.DefaultStagedMetadata, errNoApplicableMetadata
 }
 
-func (e *Entry) shouldUpdatePoliciesWithLock(currTime time.Time, sp policy.StagedPolicies) bool {
-	if e.cutoverNanos == uninitializedCutoverNanos {
-		return true
-	}
-	// If this is a future policy, we don't update the existing policy
-	// and instead use the cached policy.
-	if currTime.UnixNano() < sp.CutoverNanos {
-		e.metrics.futurePolicy.Inc(1)
-		return false
-	}
-	// If this is a stale policy, we don't update the existing policy
-	// and instead use the cached policy.
-	if sp.CutoverNanos < e.cutoverNanos {
-		e.metrics.stalePolicy.Inc(1)
-		return false
-	}
-	if e.cutoverNanos != sp.CutoverNanos {
-		return true
-	}
-	policies, useDefaultPolicies := sp.Policies()
-	if e.useDefaultPolicies && useDefaultPolicies {
+// NB: The metadata passed in is guaranteed to have cut over based on the current time.
+func (e *Entry) shouldUpdateMetadatasWithLock(
+	currTime time.Time,
+	sm metadata.StagedMetadata,
+) bool {
+	// If this is a stale metadata, we don't update the existing metadata.
+	if e.cutoverNanos > sm.CutoverNanos {
+		e.metrics.staleMetadata.Inc(1)
 		return false
 	}
-	if useDefaultPolicies {
-		policies = e.opts.DefaultPolicies()
-	}
-	return e.hasPolicyChangesWithLock(policies)
-}
 
-func (e *Entry) hasPolicyChangesWithLock(newPolicies []policy.Policy) bool {
-	if len(e.aggregations) != len(newPolicies) {
+	// If this is a newer metadata, we always update.
```

Comment:
Do we still need to check future metadata? Seems to be in the old logic?

Reply:
So I changed the logic in `activeStagedMetadataWithLock` to return an error if we can't find a staged metadata that has been cut over, and as a result, the metadata passed to `shouldUpdateMetadatasWithLock` should always have been cut over (see the NB above `shouldUpdateMetadatasWithLock`).